Scripts/WikiUpdater.cs

```diff
@@ -0,0 +1,415 @@
+            .Replace("<b>", "[b]")
+            .Replace("</b>", "[/b]")
+            .Replace("<i>", "[i]")
+            .Replace("</i>", "[/i]")
+            .Replace("\"", "\\\"");
+    }
+
+    /// <summary>
+    ///   Inserts into en.po the English translations for all the translation keys in a list of wiki pages.
+    /// </summary>
+    private static async Task InsertTranslatedPageContent(List<TranslationPair> pages,
+        CancellationToken cancellationToken)
+    {
+        // Create the whole list of values to replace first, then replace asynchronously based on read lines
+        var translationPairs = new Dictionary<string, string>();
+
+        foreach (var page in pages)
+        {
+            var untranslatedPage = page.UntranslatedPage;
+            var translatedPage = page.TranslatedPage;
+
+            // Translate page names
+            translationPairs.Add(untranslatedPage.Name, translatedPage.Name);
+
+            for (var j = 0; j < untranslatedPage.Sections.Count; j++)
+            {
+                var untranslatedSection = untranslatedPage.Sections[j];
+                var translatedSection = translatedPage.Sections[j];
+
+                // Translate body sections
+                translationPairs.Add(untranslatedSection.SectionBody, translatedSection.SectionBody);
+
+                if (untranslatedSection.SectionHeading != null && translatedSection.SectionHeading != null)
+                {
+                    // Translate headings if present and not already in the list to translate
+                    translationPairs.TryAdd(untranslatedSection.SectionHeading, translatedSection.SectionHeading);
+                }
+            }
+        }
+
+        var reader = File.ReadLinesAsync(ENGLISH_TRANSLATION_FILE, Encoding.UTF8, cancellationToken);
+        var writer = new StreamWriter(File.Create(TEMP_TRANSLATION_FILE), new UTF8Encoding(false));
+
+        var isReplacingValue = false;
+        var isFuzzyValue = false;
+        await foreach (var line in reader)
+        {
+            if (line == string.Empty)
+            {
+                // Blank lines mark the end of a value we might be replacing, so stop replacing
+                isReplacingValue = false;
+                await writer.WriteLineAsync(line);
+                continue;
+            }
+
+            if (isReplacingValue)
+            {
+                // Skip lines for values we're replacing
+                continue;
+            }
+
+            if (line == "#, fuzzy")
+            {
+                // Defer adding fuzzy labels so we can include/exclude them based on the next line
+                isFuzzyValue = true;
+                continue;
+            }
+
+            if (!line.Contains("msgid \""))
+            {
+                // Copy existing lines that we don't want to replace
+                await writer.WriteLineAsync(line);
+                continue;
+            }
+
+            var key = line.Split("\"")[1];
```

Comment:
This isn't the approach I'd use but this probably works in all cases I can think of. Anyway reading through this code once it seems straightforward that I can probably bugfix this years down the road if necessary.

Reply:
I would have used a regex to detect and extract the msgid (as I've written this kind of .po file parsing before for the check scripts) and I would have basically structured a loop to read the file line by line and copy it to the output unless the regex matched. In that case I would have omitted the previous content and written the new data to the file before continuing to loop. It's almost the same as your approach (though I see you handled the fuzzy corner cases, so in the end my simple approach would probably have gotten a bit more complicated when it was hit with real data). What I would have ended up writing would have been pretty close to what you wrote, but I would have used a regex to ensure that the msgid extraction isn't able to go wrong and accidentally process like a commented out line with msgid text on it (which I think your script might be able to hit and get into an incorrect state).

Reply:
Ah true, regex would be be more robust than contains. I’ll fix that this evening. Otherwise though I don’t think there’s a real difference in these approaches.